src/Hosting/PageLocationExpander.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2014-2020 Sarin Na Wangkanai, All Rights Reserved.
+// The Apache v2. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Razor;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+
+namespace Wangkanai.Detection.Hosting
+{
+    public class ResponsivePageLocationExpander : IViewLocationExpander
+    {
+        private const string ValueKey = "device";
+
+        public IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context, IEnumerable<string> viewLocations)
+        {
+            context.Values.TryGetValue(ValueKey, out var device);
+
+            if (!(context.ActionContext.ActionDescriptor is PageActionDescriptor))
+                return viewLocations;
+
+            if (string.IsNullOrEmpty(context.PageName) || string.IsNullOrEmpty(device))
+                return viewLocations;
+
+            var expandLocations = ExpandPageHierarchy().ToList();
+            return expandLocations;
+
+            IEnumerable<string> ExpandPageHierarchy()
+            {
+                foreach (var location in viewLocations)
+                {
+                    if (!location.Contains("/{1}/") && !location.Contains("/Shared/") || location.Contains("/Views/"))
+                    {
+                        yield return location;
+                        continue;
+                    }
+
+                    // Device View if exist on disk
+                    yield return location.Replace("{0}", "{0}." + device);
+                    // Fallback to the original default view
+                    yield return location;
+                }
```

Comment:
Did you manage to figure this out? Anything I can help with still?

Reply:
Please please help! I have my head spinning for a few days on this already. Much appreciated :-)

Reply:
OK. I'll take a look tomorrow.

Reply:
Thank you very much :-)

Reply:
I think I see what's going on here.

We don't use view location expanders to locate the "view" for a Razor Page because we've already found it. The page itself is the "view" so we don't go through a lookup process. What you could do instead is use routing a select a different page.

The other problem that's solved by doing this selection in the routing layer is one you might not have noticed. Because you defined a page with the filename `Index.mobile.cshtml` that means that it will respond to the url `/Index.mobile` which is probably not what you wanted. Routing for pages is customizable to address all of this.

![image](https://user-images.githubusercontent.com/1430011/74215173-0b13ce80-4c55-11ea-865e-f9a4940497a9.png)
![image](https://user-images.githubusercontent.com/1430011/74215182-15ce6380-4c55-11ea-8777-45899d409220.png)

I've got a branch where I've coded this up. I'll send you a link, and I'm happy to answer whatever questions you have about it.

Reply:
https://github.com/rynowak/Detection/commit/692377f3518c4c0f9fffd92caf7b7f95b1f6f215

Reply:
> [rynowak@692377f](https://github.com/rynowak/Detection/commit/692377f3518c4c0f9fffd92caf7b7f95b1f6f215)

This code awesome, its open my perspective.
I see you a create `ResponsiveAttribute`, but see you use anywhere?

Reply:
It's the convention for metadata types to use an attribute type. This would work for controllers as well. So if you had different action methods based on the device in use, that would "just work" with a controller as well.

Reply:
Thank you very much for the explanation. 

This also me wanna to revisit another issue https://github.com/wangkanai/Detection/issues/243 Responsive Option to include Web Api Controllers. By using conventions pattern might be a better way then to pass in options.